### PR TITLE
Test bench for tag expansion script

### PR DESCRIPTION
**Pull request by piperlincoln**
There are four new tests in `Testing/Tests/tag_expansion_test.py` as well as one new sample data file under `Testing/SampleData` for `TagExpansion.py`. In addition, a separate cleanup function was added to `Testing/Tests/graveyard_removal_test.py` to remove files written to disk while testing.

Finally, `TagExpansion.py` has been updated to eliminate the need for two moab instances while expanding vector tags, based on the update to Moab 5.1.0. (Specified in #34).

### CODE REVIEW

**Comment by gonuke** [comment]
I know this is only about adding tests, but I think creating tests exposes some design questions that are similar to the graveyard removal work.  You might want to rethink how to get better separation of concerns around input/output, processing of tags, filename creation, etc.

File: PythonTool/TagExpansion.py [hunk 1]
```diff
@@ -114,20 +114,51 @@ def get_tag_lists(mb, element_type, element_id):
...
+    Returns:
+    ________
+       dir_name: str
+           The name of the directory in which to create and populate each
+           vector tag directory.
+    """
+
+    # Create a directory for the vector tag expansion files.
+    if main_dir_name is None:
+        input_list = mesh_file.split("/")
```
Comment (kkiesling) [inline PythonTool/TagExpansion.py]: It appears that `mesh_file` is only passed here for the sake of naming the new files created directory, but there is also an optional `main_dir_name` argument. Could this be rolled into 1 single argument called `output_dir`? The directory created can have the same root name as the files generated, so there is no real reason they have to be two separate arguments.

If anything you could add an optional boolean argument for overwriting files already present. For example, `True` to overwrite a directory that already exists or `False` to make a new directory (appended w/ a number or so).

File: PythonTool/TagExpansion.py [hunk 2]
```diff
@@ -114,20 +114,51 @@ def get_tag_lists(mb, element_type, element_id):
     return element_list, scalar_tags, vector_tags
 
 
-def create_database(mesh_file, mb_ref, mb_exp, elements, scal_tags, vec_tag, dir_name):
+def create_directory(mesh_file, main_dir_name = None):
```
Comment (kkiesling) [inline PythonTool/TagExpansion.py]: It appears that mesh_file is only passed here for the sake of naming the new files created directory, but there is also an optional main_dir_name argument. Could this be rolled into 1 single argument called output_dir? The directory created can have the same root name as the files generated, so there is no real reason they have to be two separate arguments.

If anything you could add an optional boolean argument for overwriting files already present. For example, True to overwrite a directory that already exists or False to make a new directory (appended w/ a number or so).

File: Testing/Tests/tag_expansion_test.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+mb = pymoab.core.Core()
+mb.load_file(data_file)
+
+
+def test_get_tag_lists():
+    """
+    Ensure this function correctly retrieves all vector tags on an element.
+    """
+    elements, scalars, vectors = get_tag_lists(mb, "hex", pymoab.types.MBHEX)
+    assert len(vectors) == 2
```
Comment (kkiesling) [inline Testing/Tests/tag_expansion_test.py]: do you need to test anything about the `elements` or `scalars` that are returned here too?

File: Testing/Tests/tag_expansion_test.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    """
+    with pytest.raises(LookupError) as e:
+        get_tag_lists(mb, "tet", pymoab.types.MBTET)
+
+
+def test_create_directory():
+    """
+    Ensure this function correctly creates a directory for tag expansion.
+    """
+    dir_name = create_directory(data_file, "Testing")
```
Comment (kkiesling) [inline Testing/Tests/tag_expansion_test.py]: This only tests when you supply the second argument, but there should also be a test for when that argument is left the default `None`.

File: Testing/Tests/tag_expansion_test.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    assert os.path.isdir("Testing_database") == True
```
Comment (kkiesling) [inline Testing/Tests/tag_expansion_test.py]: Do you need to remove this directory after the assertion so it doesn't dirty the repo?

File: Testing/Tests/tag_expansion_test.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    Ensure this function correctly creates a scalar tag database for a given vector tag.
+    """
+    elements = mb.get_entities_by_type(mb.get_root_set(), pymoab.types.MBHEX)
+    tag_list = mb.tag_get_tags_on_entity(elements[0])
+    scal_tags = [tag_list[0], tag_list[3], tag_list[4]]
+    vec_tag = tag_list[1]
+    os.mkdir("photon_flux_mesh_database")
+    expand_vec_tag(mb, elements, scal_tags, vec_tag, "photon_flux_mesh_database")
+    path, dirs, files = next(os.walk("photon_flux_mesh_database/photon_result_database"))
+    assert len(files) == 24
```
Comment (kkiesling) [inline Testing/Tests/tag_expansion_test.py]: I am not sure that just checking the number of files created is correct way to go because this doesn't actually test if the tag expansion properly transferred the data. I think you should also use a file comparison to make sure the files write are correct. I think they are VTK files so using `filecmp` package should work.

File: Testing/Tests/tag_expansion_test.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+def test_tag_expansion():
+    """
+    Ensure that TagExpansion correctly expands all vector tags in a mesh.
+    """
+    os.system("python PythonTool/TagExpansion.py %s" % data_file)
+    path, dirs, files = next(os.walk("photon_flux_mesh_database/photon_result_database"))
+    assert len(files) == 24
+
+
+def test_cleanup():
```
Comment (kkiesling) [inline Testing/Tests/tag_expansion_test.py:67]: This doesn't test anything in the script that you wrote. If the purpose is to remove the files generated by a test, then these lines should be moved to inside the tests where the files are generated after the assertions are made.

**Review by kkiesling — CHANGES_REQUESTED** [review]
Thanks for adding these tests @piperlincoln!

You do a pretty good job testing function by function. However, I think you need a few more tests and assertions in your tests. These tests seem to focus on testing that the number of files are correct or that the lengths are correct. But tests need to also check that values are correctly generated. In this case, the tests should include assertions for checking the file contents as well. Also, for any test that has optional arguments, make sure that you have tests that include the optional arguments and tests that don't include them (or use the default value) so that all the behavior is being tested.

File: Testing/Tests/tag_expansion_test.py (hunk 7 above)
Comment (piperlincoln) [reply Testing/Tests/tag_expansion_test.py:67]: I created this because of a previous code review in which Baptiste and Jordan recommended that I clean up after my tests in a separate function that will always run. If the clean up statements are moved to inside the test, if the test fails, they won't be run and the repo will be dirty. When they are in their own function like this, the repo always gets cleaned no matter if any test failed or not!

Comment (gonuke) [reply Testing/Tests/tag_expansion_test.py:67]: you might want to give it a name that doesn't include `test` so that it doesn't get run as another test by the test harness

Comment (piperlincoln) [reply Testing/Tests/tag_expansion_test.py:67]: I chose to name it this way so that it will always get run no matter what. If I name it something without `test`, I would need to call it from inside one of the test functions, which introduces the same problem I was trying to avoid of the clean up not actually happening if the test fails!

File: PythonTool/TagExpansion.py [hunk 8]
```diff
@@ -160,97 +191,56 @@ def create_database(mesh_file, mb_ref, mb_exp, elements, scal_tags, vec_tag, dir
...
-    for tag in vec_tags_ref:
-        mb_ref.tag_delete(tag)
+    # Load the mesh file.
+    mb = core.Core()
+    mb.load_file(args.meshfile)
 
-    # Load the mesh file for extracting vector tag data.
-    mb_exp = core.Core()
-    mb_exp.load_file(mesh_file)
+    # Warn the user if the MB element type is not valid.
```
Comment (gonuke) [inline PythonTool/TagExpansion.py]: Isn't this taken care of by `argparse`?

File: PythonTool/TagExpansion.py [hunk 9]
```diff
@@ -160,97 +191,56 @@ def create_database(mesh_file, mb_ref, mb_exp, elements, scal_tags, vec_tag, dir
...
+    try:
+        global elements
+        mb_type = elements[args.element]
+    except KeyError as e:
+        sys.exit("WARNING: The mesh file did not contain any {} elements.".format(args.element))
 
     # Retrieve the lists of scalar and vector tags on the mesh.
-    elements_exp, scal_tags_exp, vec_tags_exp = get_tag_lists(mb_exp, element_type, mb_type)
+    try:
+        elements, scal_tags, vec_tags = get_tag_lists(mb, args.element, mb_type)
```
Comment (gonuke) [inline PythonTool/TagExpansion.py]: It looks like `mb_type` is only used here, so maybe you can define it within `get_tag_lists`?

File: PythonTool/TagExpansion.py [hunk 10]
```diff
@@ -160,97 +191,56 @@ def create_database(mesh_file, mb_ref, mb_exp, elements, scal_tags, vec_tag, dir
...
-    # Create a directory for the vector tag expansion files.
-    if main_dir_name is None:
-        input_list = mesh_file.split("/")
-        file_name = '.'.join(input_list[-1].split(".")[:-1])
-        dir_name = file_name + "_database"
-    else:
-        dir_name = main_dir_name + "_database"
+    # Warn the user if the mesh file does not contain at least one vector tag.
+    if len(vec_tags) < 1:
+        sys.exit("WARNING: This mesh file did not contain any vector tags on {} elements.".format(element_type))
```
Comment (gonuke) [inline PythonTool/TagExpansion.py]: Do you want to `raise` an exception instead of the exit?

File: PythonTool/TagExpansion.py (hunk 2 above)
Comment (gonuke) [reply PythonTool/TagExpansion.py]: I also recommend that you keep the action of `mkdir` separate from the work of determining the dir name.  Return the dirname and call `mkdir` from the calling method.

File: PythonTool/TagExpansion.py [hunk 11]
```diff
@@ -114,20 +114,51 @@ def get_tag_lists(mb, element_type, element_id):
...
+    ________
+       dir_name: str
+           The name of the directory in which to create and populate each
+           vector tag directory.
+    """
+
+    # Create a directory for the vector tag expansion files.
+    if main_dir_name is None:
+        input_list = mesh_file.split("/")
+        file_name = '.'.join(input_list[-1].split(".")[:-1])
```
Comment (gonuke) [inline PythonTool/TagExpansion.py]:
```suggestion
        main_dir_name = '.'.join(input_list[-1].split(".")[:-1])
```

File: PythonTool/TagExpansion.py [hunk 12]
```diff
@@ -114,20 +114,51 @@ def get_tag_lists(mb, element_type, element_id):
...
+        dir_name = file_name + "_database"
```
Comment (gonuke) [inline PythonTool/TagExpansion.py]: With the above change, then you don't need the else - and always use line #141

File: PythonTool/TagExpansion.py [hunk 13]
```diff
@@ -160,97 +191,56 @@ def create_database(mesh_file, mb_ref, mb_exp, elements, scal_tags, vec_tag, dir
     index = 0
     while index < length:
```
Comment (gonuke) [inline PythonTool/TagExpansion.py]: Why not a `for` loop?

File: PythonTool/TagExpansion.py [hunk 14]
```diff
@@ -160,97 +191,56 @@ def create_database(mesh_file, mb_ref, mb_exp, elements, scal_tags, vec_tag, dir
...
         scalar_data = []
-        data = mb_exp.tag_get_data(vec_tag, elements)
+        data = mb.tag_get_data(vec_tag, elements)
```
Comment (gonuke) [inline PythonTool/TagExpansion.py]: Shouldn't you get this data once, outside the loop?

File: PythonTool/TagExpansion.py [hunk 15]
```diff
@@ -160,97 +191,56 @@ def create_database(mesh_file, mb_ref, mb_exp, elements, scal_tags, vec_tag, dir
...
         scalar_data = np.copy(data[:,index])
         data_type = vec_tag.get_data_type()
```
Comment (gonuke) [inline PythonTool/TagExpansion.py]: Can't this happen outside the loop?

File: PythonTool/TagExpansion.py [hunk 16]
```diff
@@ -160,97 +191,56 @@ def create_database(mesh_file, mb_ref, mb_exp, elements, scal_tags, vec_tag, dir
...
-        scalar_tag = mb_ref.tag_get_handle(name, 1, data_type, types.MB_TAG_SPARSE,
-                                           create_if_missing = True)
-        mb_ref.tag_set_data(scalar_tag, elements, scalar_data)
+        scalar_tag = mb.tag_get_handle(str(name + "_exp"), 1, data_type, types.MB_TAG_SPARSE,
```
Comment (gonuke) [inline PythonTool/TagExpansion.py]: This can happen outside the loop, as well

**Review by gonuke — CHANGES_REQUESTED** [review]
A few recommendations to simply some of the logic throughout

File: PythonTool/TagExpansion.py (hunk 8 above)
Comment (piperlincoln) [reply PythonTool/TagExpansion.py]: In this instance, I am accounting for the case that even if the user enters a valid MB element type, the file might not actually contain any of that element! This comment is misleading and not representative of what I am actually doing, so I will fix that.

**Review by gonuke — APPROVED** [review]

**Comment by gonuke** [comment]
Unless @kkiesling  has more comments, I think we can merge this

**Review by kkiesling — APPROVED** [review]
LGTM! Thanks @piperlincoln!
